core/libp2p/multi/multibase/multibase.cpp

```diff
@@ -0,0 +1,141 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <map>
+#include <optional>
+
+#include "libp2p/multi/multibase.hpp"
+#include "libp2p/multi/multibase/codec.hpp"
+#include "libp2p/multi/multibase/codecs/base16.hpp"
+#include "libp2p/multi/multibase/codecs/base58.hpp"
+#include "libp2p/multi/multibase/codecs/base64.hpp"
+
+namespace {
+  using namespace libp2p::multi;
+
+  /**
+   * Get the encoding by a character
+   * @param ch of encoding
+   * @return related encoding, if character stands for one of them, none
+   * otherwise
+   */
+  std::optional<Multibase::Encoding> encodingByChar(char ch) {
+    switch (ch) {
+      case 'f':
+        return Multibase::Encoding::kBase16Lower;
+      case 'F':
+        return Multibase::Encoding::kBase16Upper;
+      case 'Z':
+        return Multibase::Encoding::kBase58;
+      case 'm':
+        return Multibase::Encoding::kBase64;
+      default:
+        return {};
+    }
+  }
+
+  /**
+   * Get the character, standing for the specified encoding
+   * @param encoding to be converted to a char
+   * @return char
+   */
+  char charByEncoding(Multibase::Encoding encoding) {
+    switch (encoding) {
+      case Multibase::Encoding::kBase16Lower:
+        return 'f';
+      case Multibase::Encoding::kBase16Upper:
+        return 'F';
+      case Multibase::Encoding::kBase58:
+        return 'Z';
+      case Multibase::Encoding::kBase64:
+        return 'm';
+    }
+  }
+
+  // all available codecs
+  const std::map<Multibase::Encoding, std::shared_ptr<Codec>> codecs{
+      {Multibase::Encoding::kBase16Lower,
+       std::make_shared<Base16Codec<false>>()},
+      {Multibase::Encoding::kBase16Upper,
+       std::make_shared<Base16Codec<true>>()},
+      {Multibase::Encoding::kBase58, std::make_shared<Base58Codec>()},
+      {Multibase::Encoding::kBase64, std::make_shared<Base64Codec>()}};
+}  // namespace
+
+namespace libp2p::multi {
+  using namespace kagome::expected;
+
+  Multibase::FactoryResult Multibase::createMultibaseFromEncoded(
+      std::string_view encoded_data) {
+    if (encoded_data.length() < 2) {
+      return Error{"multibase must be at least 2 characters"};
+    }
+
+    auto encoding_base = encodingByChar(encoded_data.front());
+    if (!encoding_base) {
+      return Error{"base of encoding is either unsupported or does not exist"};
+    }
+
+    std::optional<ByteBuffer> decoded_data;
+    std::string error;
+    codecs.at(*encoding_base)
+        ->decode(encoded_data.substr(1))  // cut the prefix
+        .match(
+            [&decoded_data](const Value<ByteBuffer> bytes) mutable {
+              decoded_data = bytes.value;
+            },
+            [&error](const Error<std::string> err) mutable {
+              error = err.error;
+            });
+
+    if (!decoded_data) {
+      return Error{error};
+    }
+    Multibase res{
+        std::string{encoded_data}, std::move(*decoded_data), *encoding_base};
+    return Value{std::make_unique<Multibase>(std::move(res))};
```

Comment:
did you consider *not* using pointers?

and why do you need to store both encoded and decoded data?
just store encoded and return slice for data